qiskit/qasm3/exporter.py

```diff
@@ -1108,6 +1110,21 @@ def build_for_loop(self, instruction: CircuitInstruction) -> ast.ForLoopStatemen
                     end=self.build_integer(indexset.stop - 1),
                     step=self.build_integer(indexset.step) if indexset.step != 1 else None,
                 )
+            elif isinstance(indexset, DynamicRange):
+                # OpenQASM 3 uses inclusive ranges on both ends, unlike Python.
+                if isinstance(indexset.start, int):
+                    start = self.build_integer(indexset.start)
+                else:
+                    start = self.build_expression(indexset.start)
+                if isinstance(indexset.stop, int):
+                    stop = self.build_integer(indexset.stop - 1)
+                else:
+                    stop = self.build_expression(indexset.stop)
+                if isinstance(indexset.step, int):
+                    step = self.build_integer(indexset.step) if indexset.step != 1 else None
+                else:
+                    step = self.build_expression(indexset.step)
+                indexset_ast = ast.Range(start=start, end=stop, step=step)
```

Comment:
Assuming we do pivot to `DynamicRange` getting called `expr.Range` and taking all its arguments as type-checked `expr.Expr` values, this would all just simplify to `start = None if start is None else self.build_expression(start)` etc.